### ISSUE TITLE
creating TraceExporter disables OpenTelemetry if opentelemetry-opencensus-shim is in use

### ISSUE THREAD

**Issue by barend** [GoogleCloudPlatform/opentelemetry-operations-java#141]
I am working on an application that uses OpenTelemetry natively. It also uses the BigTable client, and I'm looking at the opentelemetry-opencensus-shim to include its tracing information.

I ran into an initialisation-order problem while integrating the GCP Trace Exporter for OpenTelemetry. A workaround exists.

### Classpath

These are simply the up-to-date versions of the libraries involved. The one thing worth noting here is that this forces a newer version of the opencensus-api than requested by the GCP SDK packages. This is required to enable the opentelemetry-opencensus-shim.

- com.gooogle.cloud.opentelemetry:exporter-trace:0.20.0
- io.opencensus:opencensus-api version 0.31.0 (any version greater than 0.28.0)
- io.opentelemetry:opentelemetry-sdk version:1.11.0
- io.opentelemetry:opentelemetry-opencensus-shim:1.11.0-alpha

### Problem Description

If the opentelemetry-opencensus-shim is on the classpath, and the `GOOGLE_CLOUD_PROJECT` variable is not set, then constructing the `TraceConfiguration` initializes the `GlobalOpenTelemetry` singleton to `OpenTelemetry.noop()`. This causes the intended configuration of the OpenTelemetry SDK to fail with an exception. If this exception is uncaught, the application fails to start up; if it is caught, it'll work but the telemetry will be inert.

### Problem Details

This happens because of the project ID lookup in `com.google.cloud.ServiceOptions#getDefaultProjectId`, triggered by [`TraceConfiguration#DEFAULT_PROJECT_ID`][ref1]. This function early-returns if the environment variable is present. If it isn't, it initiates a http call to AppEngine, which is instrumented by OpenCensus. This causes initialization of the opencensus-shim ([source][ref2]), which in turn initializes OpenTelemetry. In absence of a configured OpenTelemetry SDK (which we're in the process of configuring), this defaults to the no-op SDK. When we finish building our desired OpenTelemetry config, the `buildAndRegisterGlobal()` call throws an IllegalStateException.

### Sample code

This is an extract of the code I use at startup.

```scala
// The <clinit> of TraceConfiguration ultimately leads to a call to GlobalOpenTelemetry.set()
val gcpTraceExporter = TraceExporter.createWithConfiguration(TraceConfiguration.builder()
    .setFixedAttributes(/* ... */)
    .build())

val otel = OpenTelemetrySdk.builder()
  .setTracerProvider(SdkTracerProvider.builder()
    .setResource(/* ... */)
    .setSampler(/* ... */)
    .addSpanProcessor(BatchSpanProcessor
      .builder(gcpTraceExporter)
      .build())
    .build()
  ).buildAndRegisterGlobal()  // kaboom
```

### Workaround

Define the `GOOGLE_CLOUD_PROJECT` environment variable or JVM system property.

### Possible Solution

Lazy initialisation of the default project ID inside `TraceConfiguration` might just do it. This problem will eventually disappear entirely when the wider GCP SDK adopts OpenTelemetry in place of OpenCensus.

[ref1]: https://github.com/GoogleCloudPlatform/opentelemetry-operations-java/blob/v0.20.0/exporters/trace/src/main/java/com/google/cloud/opentelemetry/trace/TraceConfiguration.java#L38-L39
[ref2]: https://github.com/census-instrumentation/opencensus-java/blob/v0.31.0/api/src/main/java/io/opencensus/trace/Tracing.java#L93-L106

**Comment by jsuereth** [comment]
So the only thing `TraceConfiguration` does is look up default project id using GAX.   I'm guessing the call to `GlobalOpenTelemetry.set()` is via something within GAX that attempts to pull OpenCensus and then OpenCensus attempting to load OTEL if available.

Theoretically, these should be completely disentangled, as the helper method we use shouldn't require OpenCensus instantatiation yet.   Will continue to investigate.

**Comment by dashpole** [comment]
@psx95 can you take a look at this?  The gRPC team ran into this trying to migrate from OC -> OTel.

**Comment by YifeiZhuang** [comment]
Any update on this?

**Comment by psx95** [comment]
Sorry, I was on a vacation, so no updates as of yet. 
I will look into it.

**Comment by psx95** [comment]
@YifeiZhuang I was able to reproduce the issue as described when using the dependency versions mentioned above. 

However, I ran my sample application with the same dependencies, except for updating the `com.gooogle.cloud.opentelemetry:exporter-trace` to the current HEAD (I manually built a JAR and added it as the dependency) and then my application started successfully.

The current latest version for this dependency that is published is v0.23.0 and in my tests, even that will cause application to fail. 

Can you try building a shadowJar for the exporter-trace dependency directly from the source and verify once if it works ?

**Comment by psx95** [comment]
Between the HEAD and and v0.20 for the trace exporter, the following upstream dependencies changed from v0.10.0 to v0.15.0 - 

io.opentelemetry:opentelemetry-api
io.opentelemetry:opentelemetry-context
io.opentelemetry:opentelemetry-sdk
io.opentelemetry:opentelemetry-sdk-common
io.opentelemetry:opentelemetry-sdk-extension-autoconfigure-spi
io.opentelemetry:opentelemetry-sdk-extension-autoconfigure
io.opentelemetry:opentelemetry-semconv
io.opentelemetry:opentelemetry-sdk-metrics
io.opentelemetry:opentelemetry-sdk-extension-autoconfigure
Io.opentelemetry:opentelemetry-sdk-extension-resources

I verified that simply upgrading these dependency versions in opentelemetry-java-operations code tagged at v0.20.0 (*and no other logical code change*) would resolve the issue. 

Creating a new release from HEAD should resolve this for everyone.

**Comment by psx95** [comment]
Feel free to re-open this issue if there is additional information or if the above resolution does not work for you.

**Comment by psx95** [comment]
A possible workaround for this bug that has been tested is use of [OpenTelemetry Java SDK extension - autoconfigure](https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/autoconfigure) for configuring and setting up OpenTelemetry for your application.

An example of how to use autoconfigure can be found [here](https://github.com/open-telemetry/opentelemetry-java-docs/tree/main/autoconfigure).

**Comment by barend** [comment]
Thank you for the updates. I think the workaround I posted in the original update is still in place. We still use the in-process GCP exporter, and it has been well-behaved enough that tbh I had forgotten about this ticket. We have kept up to date with our library versions, so that extra env var may technically no longer be required.

I did make one related change elsewhere, which is that I copied the [`OpencensusTracer`](https://github.com/googleapis/gax-java/blob/main/gax/src/main/java/com/google/api/gax/tracing/OpencensusTracer.java) and `-Factory` classes out of the `googleapis/gax-java` project into ours, and rewrote them to use OpenTelemetry instead of Opencensus. This let me drop the opentelemetry-opencensus-shim from our project dependencies.

OpenTelemetry was still a little rough one year ago, but it has stabilised in the meantime. By now, the best follow-up for this ticket IMHO would be for the gax-java project to offer an OpentelemetryTracer out of the box. With regards to the opentelemetry-operations-java project, I think this ticket can safely be closed.

**Comment by barend** [comment]
I realise that the preferred deployment for OpenTelemetry in general is to use the OTLP exporter to an otel Collector on localhost, which is what I would run on GKE, but we're on Cloud Run. The in-process exporter has been well-behaved all year.

**Comment by psx95** [comment]
This same issue exists with the exporter-metrics module as well. Will track that in a separate issue.

**Comment by psx95** [comment]
This issue should now be fixed by the changes in #193.
